### PR TITLE
[cli] show network parameters 782

### LINKED ISSUE #782: Power users SBAT see current network parameters via CLI

### PR DESCRIPTION
**Pull request by mrsmkl**
### Description

Prints information about parameters of the network.

### Tested

Just tested the commands on local ganache node.

### Other changes

Added some wrappers to `contractkit`.

### Related issues

- Fixes #782

### CODE REVIEW

File: packages/protocol/truffle.js [hunk 1]
```diff
@@ -19,12 +19,14 @@ const ALFAJORES_FROM = '0x456f41406B32c45D59E539e4BBA3D7898c3584dA'
 const PILOT_FROM = '0x387bCb16Bfcd37AccEcF5c9eB2938E30d3aB8BF2'
 const PILOTSTAGING_FROM = '0x545DEBe3030B570731EDab192640804AC8Cf65CA'
 
+const gasLimit = 8000000
```
Comment (mcortesi) [inline packages/protocol/truffle.js]: do we need to increase the gasLimit? 
@m-chrzan ??

File: packages/contractkit/src/wrappers/Reserve.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Reserve } from '../generated/types/Reserve'
+import { BaseWrapper, proxyCall, toBigNumber } from './BaseWrapper'
+
+export class ReserveWrapper extends BaseWrapper<Reserve> {
+  tobinTaxStalenessThreshold = proxyCall(
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Reserve.ts:17]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Governance.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { Governance } from '../generated/types/Governance'
+import { BaseWrapper, proxyCall, toBigNumber } from './BaseWrapper'
+
+export class GovernanceWrapper extends BaseWrapper<Governance> {
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Governance.ts:22]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/GasPriceMinimum.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { GasPriceMinimum } from '../generated/types/GasPriceMinimum'
+import { BaseWrapper, proxyCall, toBigNumber } from './BaseWrapper'
+
+export class GasPriceMinimumWrapper extends BaseWrapper<GasPriceMinimum> {
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/GasPriceMinimum.ts:15]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/SortedOracles.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import { SortedOracles } from '../generated/types/SortedOracles'
+import { BaseWrapper, proxyCall, toBigNumber } from './BaseWrapper'
+
+export class SortedOraclesWrapper extends BaseWrapper<SortedOracles> {
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/SortedOracles.ts:12]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Validators.ts [hunk 6]
```diff
@@ -34,6 +41,19 @@ export class ValidatorsWrapper extends BaseWrapper<Validators> {
   removeMember = proxySend(this.kit, this.contract.methods.removeMember)
   registerValidator = proxySend(this.kit, this.contract.methods.registerValidator)
   registerValidatorGroup = proxySend(this.kit, this.contract.methods.registerValidatorGroup)
+  minElectableValidators = proxyCall(
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Validators.ts:64]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Validators.ts [hunk 7]
```diff
@@ -34,6 +41,19 @@ export class ValidatorsWrapper extends BaseWrapper<Validators> {
...
+    this.contract.methods.minElectableValidators,
+    undefined,
+    toBigNumber
+  )
+  maxElectableValidators = proxyCall(
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Validators.ts:73]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Validators.ts [hunk 8]
```diff
@@ -34,6 +41,19 @@ export class ValidatorsWrapper extends BaseWrapper<Validators> {
...
+    this.contract.methods.maxElectableValidators,
+    undefined,
+    toBigNumber
+  )
+  electionThreshold = proxyCall(this.contract.methods.getElectionThreshold, undefined, toBigNumber)
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Validators.ts:82]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Validators.ts [hunk 9]
```diff
@@ -34,6 +41,19 @@ export class ValidatorsWrapper extends BaseWrapper<Validators> {
...
+  getRegistrationRequirement = proxyCall(this.contract.methods.getRegistrationRequirement)
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Validators.ts]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Exchange.ts [hunk 10]
```diff
@@ -17,6 +17,11 @@ import {
  * using a Constant Product Market Maker Model
  */
 export class ExchangeWrapper extends BaseWrapper<Exchange> {
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Exchange.ts:26]: Add [typedoc comments](https://typedoc.org/guides/doccomments/) (to all methods)

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 11]
```diff
@@ -12,6 +12,16 @@ export class AttestationsWrapper extends BaseWrapper<Attestations> {
     this.kit,
     this.contract.methods.setAccountDataEncryptionKey
   )
+  attestationExpirySeconds = proxyCall(
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Attestations.ts]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 12]
```diff
@@ -12,6 +12,16 @@ export class AttestationsWrapper extends BaseWrapper<Attestations> {
...
+    this.contract.methods.attestationExpirySeconds,
+    undefined,
+    toBigNumber
+  )
+  attestationRequestFees = proxyCall(
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Attestations.ts:112]: Add [typedoc comments](https://typedoc.org/guides/doccomments/)

File: packages/cli/src/commands/network/parameters.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+  static flags = {
+    ...BaseCommand.flags,
+  }
+
+  async run() {
+    const goldToken = await this.kit.contracts.getGoldToken()
+    const stableToken = await this.kit.contracts.getStableToken()
+    const attestations = await this.kit.contracts.getAttestations()
+    console.log(
```
Comment (mcortesi) [inline packages/cli/src/commands/network/parameters.ts]: please, investigate using some of the `cli` output function, instead of console log. 

For example, we do:
```ts
    const balances = {
      goldBalance: await goldToken.balanceOf(args.account),
      dollarBalance: await stableToken.balanceOf(args.account),
    }
    printValueMap(balances)
```

File: packages/cli/src/commands/network/parameters.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
+import { BaseCommand } from '../../base'
+
+export default class Parameters extends BaseCommand {
+  static description = 'View network parameters'
+
+  static flags = {
+    ...BaseCommand.flags,
+  }
+
+  async run() {
```
Comment (mcortesi) [inline packages/cli/src/commands/network/parameters.ts:11]: We are doing **many calls** here. We could easily run many in parallell, using something like `Promise.all()`

File: packages/cli/src/commands/network/parameters.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+    const governance = await this.kit.contracts.getGovernance()
+    console.log(
+      'governance.concurrentProposals',
+      (await governance.concurrentProposals()).toString()
+    )
+    console.log('governance.dequeueFrequency', (await governance.dequeueFrequency()).toString())
+    console.log('governance.minDeposit', (await governance.minDeposit()).toString())
+    console.log('governance.queueExpiry', (await governance.queueExpiry()).toString())
+    const durations = await governance.stageDurations()
```
Comment (mcortesi) [inline packages/cli/src/commands/network/parameters.ts]: Instead of parsing the output here. We should make `governance.stageDuraation()` return an object with `{approvalStage, referendumStage, executionStage}`. Use the postParse param from proxyCall

File: packages/cli/src/commands/network/parameters.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+      'seconds'
+    )
+
+    const sortedOracles = await this.kit.contracts.getSortedOracles()
+    console.log(
+      'sortedOracles.reportExpirySeconds',
+      (await sortedOracles.reportExpirySeconds()).toString()
+    )
+
+    const exchange = await this.kit.contracts.getExchange()
```
Comment (mcortesi) [inline packages/cli/src/commands/network/parameters.ts]: what about adding a method to the exchange wrapper called: `getConfiguration(): Promise<ExchangeConfig>` that returns all this?

The same applies to all other Contract wrappers.

File: packages/cli/src/commands/network/parameters.ts (hunk 15 above)
Comment (mcortesi) [reply packages/cli/src/commands/network/parameters.ts]: and maybe call it `getStageDurations()`

File: packages/cli/src/commands/network/parameters.ts (hunk 16 above)
Comment (mcortesi) [reply packages/cli/src/commands/network/parameters.ts]: In fact, we can do even better.

```ts
kit.getCeloConfiguration(): Promise<CeloConfiguration>

interface CeloConfig: {
    exchange: ExchangeConfig,
    .... all other contracts...
}
```

File: packages/protocol/scripts/devchain.ts [hunk 17]
```diff
@@ -63,7 +63,7 @@ async function startGanache(datadir: string, opts: { verbose?: boolean }) {
     network_id: 1101,
     db_path: datadir,
     mnemonic: MNEMONIC,
-    gasLimit: 7000000,
+    gasLimit: 8000000,
```
Comment (mcortesi) [inline packages/protocol/scripts/devchain.ts]: check `contractKit` test setup, which also runs ganache

File: packages/protocol/contracts/governance/Validators.sol [hunk 18]
```diff
@@ -60,13 +60,19 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
   LockedGoldCommitment private registrationRequirement;
   uint256 public minElectableValidators;
   uint256 public maxElectableValidators;
+  FixidityLib.Fraction electionThreshold;
```
Comment (mcortesi) [inline packages/protocol/contracts/governance/Validators.sol]: All changes here seem to be unrelated to the PR

**Review by mcortesi — CHANGES_REQUESTED** [review]

File: packages/contractkit/src/wrappers/Reserve.ts (hunk 2 above)
Comment (mrsmkl) [reply packages/contractkit/src/wrappers/Reserve.ts:17]: The wrappers don't seem to have much comments  (perhaps because many would be the same as in the contract), but I'll at least add short descriptions of each contract.

Comment (mcortesi) [reply packages/contractkit/src/wrappers/Reserve.ts:17]: You can copy docs from the contract. But follow typedoc conventions

There's a branch where i've added docs for all methods on the wrappers

Comment (mcortesi) [reply packages/contractkit/src/wrappers/Reserve.ts:17]: The idea is that we will generate docs from the .ts files and expose them in a public page.

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 19]
```diff
@@ -1,17 +1,64 @@
...
+    this.contract.methods.attestationRequestFees,
+    undefined,
+    toBigNumber
+  )
+
+  async getConfig(): Promise<AttestationsConfig> {
+    const evs = await this.contract.getPastEvents('AttestationRequestFeeSet', { fromBlock: 0 })
```
Comment (m-chrzan) [inline packages/contractkit/src/wrappers/Attestations.ts]: Prefer not shortening variable name, `events`.

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 20]
```diff
@@ -1,17 +1,64 @@
...
+    const tokenMap: any = {}
+    const tokens: string[] = []
+    evs.forEach((el) => {
```
Comment (m-chrzan) [inline packages/contractkit/src/wrappers/Attestations.ts]: Ditto, prefer `element`. Or even better, `event`.

File: packages/protocol/contracts/governance/Validators.sol [hunk 21]
```diff
@@ -483,6 +520,7 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
     require(voters[account] == address(0));
     uint256 weight = getAccountWeight(account);
     require(weight > 0);
+    totalVotes = totalVotes.add(weight);
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/Validators.sol]: Looks like you branched off of your electoral threshold PR, adding some extraneous diffs to this one.

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 22]
```diff
@@ -1,17 +1,64 @@
...
+    toBigNumber
+  )
+  attestationRequestFees = proxyCall(
+    this.contract.methods.attestationRequestFees,
+    undefined,
+    toBigNumber
+  )
+
+  async getConfig(): Promise<AttestationsConfig> {
+    const events = await this.contract.getPastEvents('AttestationRequestFeeSet', { fromBlock: 0 })
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Attestations.ts]: we are asking the node to go over the whole blockchain history to search for Events. That can be a **really heave workload** for them!

Comment (m-chrzan) [reply packages/contractkit/src/wrappers/Attestations.ts]: Agreed. Feel free to add a list of tokens to Attestations, or just use the address of the only stable token we support right now.

Also, looks like we don't actually set these fees anywhere on our deployed contracts. @nambrot, do you know anything about this? Might make sense to just leave this as a TODO for now.

File: packages/protocol/truffle.js (hunk 1 above)
Comment (m-chrzan) [reply packages/protocol/truffle.js]: Probably. Smart contracts have increased in bytecode so might need to pay a higher fee for deploying them.

File: packages/contractkit/src/wrappers/Reserve.ts (hunk 2 above)
Comment (mrsmkl) [reply packages/contractkit/src/wrappers/Reserve.ts:17]: Tried to add comments to all wrappers that I added.

File: packages/contractkit/src/kit.ts [hunk 23]
```diff
@@ -32,6 +53,44 @@ export class ContractKit {
     this.contracts = new WrapperCache(this)
   }
 
+  async getNetworkConfig(): Promise<NetworkConfig> {
+    const token1 = await this.contracts.getGoldToken()
```
Comment (mcortesi) [inline packages/contractkit/src/kit.ts]: you can use `this.register.getAddressFor()`

File: packages/contractkit/src/wrappers/Attestations.ts [hunk 24]
```diff
@@ -4,14 +4,35 @@ import BigNumber from 'bignumber.js'
...
+} from './BaseWrapper'
 const parseSignature = SignatureUtils.parseSignature
 
 export interface AttestationStat {
   completed: number
   total: number
 }
 
+export interface AttestationsToken {
+  address: string
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Attestations.ts]: there's a type alias `Address` that you can use here

File: packages/contractkit/src/wrappers/Governance.ts [hunk 25]
```diff
@@ -0,0 +1,75 @@
+import BigNumber from 'bignumber.js'
+import { Governance } from '../generated/types/Governance'
+import { BaseWrapper, proxyCall, toBigNumber } from './BaseWrapper'
+
+export interface StageDurations {
+  approval: BigNumber
```
Comment (mcortesi) [inline packages/contractkit/src/wrappers/Governance.ts]: comments if this is seconds/milliseconds/etc

**Comment by mcortesi** [comment]
I really like how this is looking like now! Thanks for that @mrsmkl 

I left a few comments, and there a couple of old ones unresolved. (the other's i've already marked them as resolved)

In particular:
 - we can't do `getPastEvents` from block 0
 - split PRs between networkConfig and other stuff

Also, not sure if you've already done this. but upgrade with master.. there were some changes in contractkit in the past weeks

**Comment by mrsmkl** [comment]
The other PR should be merged already, so there should be no changes related to that anymore.
The part where I query events is not used currently. Anyway, the events are indexed, so it shouldn't be that inefficient, unless the values are changed frequently. Don't know about the light client though. So probably best to add the efficient version as TODO, and decide whether to keep this one for now or just remove it.

**Comment by m-chrzan** [comment]
I would second Mariano's request to remove relying on events. The fact that the list of tokens for which we have a fee feels like an oversight from the past - this is on-chain information that we should have available. 

I suggest removing the logic for getting attestation request fees for now (as per my comment, it looks like we're not even setting them right now), and add an issue for having a list of tokens for which this fee can be set (maybe not necessary, potentially could reuse gas whitelist contract?).

**Review by mcortesi — APPROVED** [review]

**Comment by mcortesi** [comment]
Great! i see you removed the method ;)
